client/client.go

```diff
@@ -386,6 +386,60 @@ func (c *Client) BalanceProofWatcher() {
 	}
 }
 
+// FrozenWatcher listenes for Frozen events and calls WithdrawFrozen
+// when a balance proof is available.
+func (c *Client) FrozenWatcher() {
+	ctx, cancel := context.WithCancel(c.Ctx())
+	defer cancel()
+	sub, err := c.ethClient.SubscribeFrozen(ctx, c.contract, nil)
+	if err != nil {
+		c.logError("Frozen event subscription: %v", err)
+		return
+	}
+	defer sub.Unsubscribe()
+
+	select {
+	case err := <-sub.Err():
+		c.logError("Frozen event subscription: %v", err)
+		return
+	case event := <-sub.Events():
+		c.handleFrozen(event)
+	}
+}
+
+func (c *Client) handleFrozen(event *bindings.ErdstallFrozen) {
+	epoch := event.Epoch
+	c.events <- &Event{Type: SET_OP_TRUST, OpTrust: UNTRUSTED}
+	c.log("❄️ Contract Frozen in epoch #%d", epoch)
+
+	c.balMtx.Lock()
+	defer c.balMtx.Unlock()
+	bal, ok := c.balances[epoch]
+	if !ok {
+		c.logOffChain("No balance-proof available for freeze")
+		return
+	}
+	status := make(chan *CmdStatus)
+	defer close(status)
+	go func() {
+		for msg := range status {
+			if msg.Err != nil {
+				c.logOnChain(msg.Err.Error())
+			} else {
+				c.logOnChain(msg.Msg)
+			}
+		}
+	}()
+	_, err := c.sendTx("WithdrawFrozen", func(auth *bind.TransactOpts) (*types.Transaction, error) {
+		return c.contract.WithdrawFrozen(auth, bindings.ErdstallBalance{Epoch: epoch, Account: bal.Account, Value: bal.Value}, bal.Bal.Sig)
+	}, status)
```

Comment:
Why do you even pass a `status` channel to `sendTx` instead of just using `c.logOnChain` in it? It looks funny that you start above this logger go routine even though you could just call the `c.logOnChain` methods directly where you write `CmdStatus`'s into the channel.

Reply:
btw you're missing a newline above the def of method `sendTx`.

Reply:
looking at the code further, it seems that you always pass the `status` channel down the stack?

Reply:
There are several output windows for this, the `status` is displayed in the lower right, which tracks the progress/status of a command whereas the `log` calls are not associated to a specific command.

Reply:
The commit looks a bit weird but in the end result there is a newline above the `sendTx`.